internal/events/aggregator_batch_state.go

```diff
@@ -0,0 +1,387 @@
+// Copyright © 2022 Kaleido, Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package events
+
+import (
+	"context"
+	"crypto/sha256"
+	"database/sql/driver"
+	"encoding/binary"
+
+	"github.com/hyperledger/firefly/internal/definitions"
+	"github.com/hyperledger/firefly/internal/log"
+	"github.com/hyperledger/firefly/pkg/database"
+	"github.com/hyperledger/firefly/pkg/fftypes"
+)
+
+func newBatchState(ag *aggregator) *batchState {
+	return &batchState{
+		database:           ag.database,
+		definitions:        ag.definitions,
+		maskedContexts:     make(map[fftypes.Bytes32]*nextPinGroupState),
+		unmaskedContexts:   make(map[fftypes.Bytes32]*contextState),
+		dispatchedMessages: make([]*dispatchedMessage, 0),
+
+		PreFinalize: make([]func(ctx context.Context) error, 0),
+		Finalize:    make([]func(ctx context.Context) error, 0),
+	}
+}
+
+// nextPinGroupState manages the state during the batch for an individual masked context.
+// We read it from the database the first time a context is touched within a batch, and then
+// replace the pins in it as they are spent/replaced for each pin processed.
+// At the end we flush the changes to the database
+type nextPinGroupState struct {
+	groupID           *fftypes.Bytes32
+	topic             string
+	nextPins          []*fftypes.NextPin
+	new               bool
+	identitiesChanged map[string]bool
+}
+
+type NextPinState struct {
```

Comment:
Any reason this one is public/exported? Seems to be internal to the package.

Reply:
I was thinking of the boundary of the BatchState and its interface with Aggregator.

Agree now looking back it's confusing, as these aren't on the external interface. So while this doesn't _actually_ make them public outside the package, it is confusing.